File: 2024/day2/py/day2.py
#!/usr/bin/env python3

from sys import argv
from typing import TypeAlias

matrix: TypeAlias = list[list[int]]


def check_safety(row: list[int]) -> bool:
    deltas = [b - a for a, b in zip(row, row[1:])]

    # rule 1: increasing or decreasing slope
    slope = all(d > 0 for d in deltas) or all(d < 0 for d in deltas)

    # rule 2: bounded distance
    rate = all(3 >= abs(d) >= 1 for d in deltas)

    return slope and rate


def part1(data: matrix) -> int:
    return sum(check_safety(row) for row in data)


def part2(data: matrix) -> int:
    safe = 0
    for row in data:
        if check_safety(row):
            safe += 1
            continue

        safe += any(
            check_safety(row[:i] + row[i + 1 :]) for i in range(len(row))
        )

    return safe


def parse_data(filename: str) -> matrix:
    data: matrix = []
    with open(filename, "r") as f:
        for line in f:
            row = [int(x) for x in line.strip().split()]
            data.append(row)
    return data


def main():
    filename = "../test.txt"
    if len(argv) > 1:
        filename = argv[1]

    data = parse_data(filename)
    print(f"Part 1: {part1(data)}")
    print(f"Part 2: {part2(data)}")


if __name__ == "__main__":
    main()

File: 2024/day2/elixir/lib/day2.ex
defmodule Day2 do
  @spec parse_input(String.t()) :: [[integer()]]
  def parse_input(filename) do
    filename
    |> File.read!()
    |> String.split(~r{\n}, trim: true)
    |> Enum.map(fn line ->
      line
      |> String.split()
      |> Enum.map(&String.to_integer/1)
    end)
  end

  @doc ~s"""
  Checks if a row is safe.

  ## Examples
    
    iex> Day2.is_safe? [1,2,3,4]
    true
    iex> Day2.is_safe? [4,3,2,1]
    true
    iex> Day2.is_safe? [1,5,6,7]
    false
    iex> Day2.is_safe? [1,2,1,3]
    false
    
  """
  @spec is_safe?([integer()]) :: boolean()
  def is_safe?(row) do
    deltas =
      row
      |> Enum.chunk_every(2, 1, :discard)
      |> Enum.map(fn [a, b] -> b - a end)

    increasing_map = deltas |> Enum.all?(&Kernel.<(0, &1))
    decreasing_map = deltas |> Enum.all?(&Kernel.>(0, &1))

    bounded_map =
      deltas
      |> Enum.all?(fn a ->
        3 >= abs(a) and abs(a) >= 1
      end)

    (increasing_map or decreasing_map) and bounded_map
  end

  def part1(matrix) do
    matrix
    |> Enum.map(&if is_safe?(&1), do: 1, else: 0)
    |> Enum.sum()
  end

  def main do
    matrix = Day2.parse_input("../test.txt")

    matrix
    |> part1()
    |> IO.puts()
  end
end

Day2.main()
